File: string-calculator/csharp/src/StringCalculator/DelimiterParser.cs
using System.Text.RegularExpressions;

namespace StringCalculator;

internal static class DelimiterParser
{
    private static readonly string[] DefaultDelimiters = { ",", "\n" };

    public static (string[] Delimiters, string Body) Parse(string input)
    {
        if (!input.StartsWith("//"))
        {
            return (DefaultDelimiters, input);
        }

        var headerEnd = input.IndexOf('\n');
        var header = input.Substring(2, headerEnd - 2);
        var body = input.Substring(headerEnd + 1);

        string[] delimiters;
        if (header.StartsWith("["))
        {
            var matches = Regex.Matches(header, @"\[([^\]]+)\]");
            delimiters = new string[matches.Count];
            for (var i = 0; i < matches.Count; i++)
            {
                delimiters[i] = matches[i].Groups[1].Value;
            }
        }
        else
        {
            delimiters = new[] { header };
        }

        return (delimiters, body);
    }
}

File: string-calculator/typescript/src/delimiterParser.ts
const DEFAULT_DELIMITER = /[,\n]/;

export interface ParsedInput {
  delimiter: RegExp;
  body: string;
}

export function parse(input: string): ParsedInput {
  if (!input.startsWith('//')) {
    return { delimiter: DEFAULT_DELIMITER, body: input };
  }

  const headerEnd = input.indexOf('\n');
  const header = input.substring(2, headerEnd);
  const body = input.substring(headerEnd + 1);
  return { delimiter: new RegExp(header), body };
}
